// timber/dist/log.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _augment = require('./utils/augment');

var _augment2 = _interopRequireDefault(_augment);

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var loggers = {
  console: {
    detect: function detect() {
      return _config2.default.logger.constructor.name === 'Console' || _config2.default.logger.constructor.name === 'CustomConsole';
    },
    handler: function handler(level, message, metadata) {
      if (metadata) {
        return _config2.default.logger[level](new _augment2.default(message, metadata));
      }
      return _config2.default.logger[level](message);
    }
  },
  winston: {
    detect: function detect() {
      return _config2.default.logger.Container && _config2.default.logger.Logger && _config2.default.logger.Transport;
    },
    handler: function handler(level, message) {
      var metadata = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      return _config2.default.logger.log(level, message, metadata);
    }
  },
  bunyan: {
    detect: function detect() {
      return _config2.default.logger.constructor.name === 'Logger';
    },
    handler: function handler(level, message, metadata) {
      _config2.default.logger[level](metadata, message);
    }
  }
};

var log = function log() {
  // Iterate through the loggers object to detect
  // which logger is set in the timber config.
  for (var name in loggers) {
    // If we successfully detected the logger...
    if (loggers[name].detect()) {
      var _loggers$name;

      // Pass the provded arguments to the logger
      return (_loggers$name = loggers[name]).handler.apply(_loggers$name, arguments);
    }
  }
};

exports.default = log;
